### DISCUSSION TITLE
v0.14.0 - The International Languages Update

### DISCUSSION THREAD

**Discussion by tankerkiller125** [sysadminsmedia/homebox#184] [Announcements]
> [!NOTE]  
> We are aware of some users having issues with the v0.12.0 -> v0.13.0 update on docker specifically, we have so far been unable to replicate the issue and as such this version does not contain a patch for it. We want to reassure you though that we are committed to getting to the bottom of that issue and fixing it, whether that be a patch release, or something that goes in a fuller release.

## Homebox Goes International (Languages)

This update contains many fixes, and small resolutions, but it also contains a HUGE update in terms of making Homebox more accessible to international users. We've finally added i18n support to Homebox, allowing users to use Homebox in their own language. It's still early days, we've translated the majority of the main pages, and we continue working on it.

To use translations, simply open Homebox, it's really that simple, it will use whatever the default language setting your browser is set to use. And fallback to English if your preferred language isn't yet available. If your language isn't yet available, please consider [contributing](https://translate.sysadminsmedia.com/projects/homebox/frontend/). For those that have already contributed, both those named, and those that don't have your GitHub linked in Weblate (and thus don't appear in commit messages to mention) thank you so much, we couldn't possible translate Homebox into so many languages ourselves.

[![Translation status](http://translate.sysadminsmedia.com/widget/homebox/multi-auto.svg)](http://translate.sysadminsmedia.com/engage/homebox/)

Additionally, we purchased the https://homebox.software domain to give Homebox a home of it's own on the internet. Which should not only make it more findable for SEO reasons, but also make it easier to remember the link to the documentation/home page. Additionally, we setup some sub-domains to redirect to various sites such as https://git.homebox.software for Github, https://discord.homebox.software to go to our Discord server, etc.

## What's Changed
* Fix small typo in label discussion thread URL by @victorhooi in https://github.com/sysadminsmedia/homebox/pull/149
* Document search tips, cleanup documentation by @blessedbiped in https://github.com/sysadminsmedia/homebox/pull/152
* Fixes to the Tools page by @rpavlik in https://github.com/sysadminsmedia/homebox/pull/154
* Adding i18n initial support by @tankerkiller125 in https://github.com/sysadminsmedia/homebox/pull/155
* changed companyname and url to sysadminsmedia.com by @101br03k in https://github.com/sysadminsmedia/homebox/pull/158
* [LANGUAGE UPDATE] Frontend translations for Italian and German by @lukasitaly in https://github.com/sysadminsmedia/homebox/pull/170
* Adding email validator by @RomuloGatto in https://github.com/sysadminsmedia/homebox/pull/178
* fix: CSV export not including item notes by @LarssonOliver in https://github.com/sysadminsmedia/homebox/pull/180
* added the ability to add a photo in the item creation modal by @cjmielke in https://github.com/sysadminsmedia/homebox/pull/173

## New Contributors
* @victorhooi made their first contribution in https://github.com/sysadminsmedia/homebox/pull/149
* @blessedbiped made their first contribution in https://github.com/sysadminsmedia/homebox/pull/152
* @rpavlik made their first contribution in https://github.com/sysadminsmedia/homebox/pull/154
* @lukasitaly made their first contribution in https://github.com/sysadminsmedia/homebox/pull/170
* @RomuloGatto made their first contribution in https://github.com/sysadminsmedia/homebox/pull/178
* @LarssonOliver made their first contribution in https://github.com/sysadminsmedia/homebox/pull/180
* @cjmielke made their first contribution in https://github.com/sysadminsmedia/homebox/pull/173

## Translation Contributors
* @SKNTim (Chinese Traditional)
* @olsson82 (Chinese Traditional)
* @101br03k (Chinese Traditional, Swedish, French)
* @Jackxwb (Chinese Simplified)
* @SodaSyrup (Turkish)
* @N0namenull (Russian)
* @scyllaL (Russian)
* @Slydite4 (Spanish)
* @HydrelioxGitHub (French)
* @chevdor (French)
* @lukasitaly (Italian, German, French)
* @olsson82 (Swedish, German)
* @terenc3 (German)

**Full Changelog**: https://github.com/sysadminsmedia/homebox/compare/v0.13.0...v0.14.0

<hr /><em>This discussion was created from the release <a href='https://github.com/sysadminsmedia/homebox/releases/tag/v0.14.0'>v0.14.0 - The International Languages Update</a>.</em>

**Comment by Jackxwb** [comment]
Where can I switch the language? I can't find the option to switch the language
I tried the Docker version and the Windows version, and the menus are no different from the previous version
Version: 0.14.0 ~ Build: 02c0453ff3d21b1e54e9ee038287c0b372b4fa04
![image](https://github.com/user-attachments/assets/02805712-ccda-47cf-9d04-de811f77176c)

**Reply by tankerkiller125** [reply]
It should switch automatically based on your browser settings. You may need to clear the browser cache or CDN for your Homebox domain given it is a JS change. Of course this only applies if your language is supported (see the translations list in the Readme)

**Reply by Jackxwb** [reply]
> It should switch automatically based on your browser settings. You may need to clear the browser cache or CDN for your Homebox domain given it is a JS change. Of course this only applies if your language is supported (see the translations list in the Readme)

I have been unable to switch languages when using a new browser, clearing cookies, or switching to the privacy mode of the browser
![image](https://github.com/user-attachments/assets/c50f78fd-379f-4066-b7ea-2b9d4a11b707)
'zh-CN' corresponds to Simplified Chinese language

**Reply by Jackxwb** [reply]
> It should switch automatically based on your browser settings. You may need to clear the browser cache or CDN for your Homebox domain given it is a JS change. Of course this only applies if your language is supported (see the translations list in the Readme)

After clearing the cookies, these contents were generated and no language related configurations were seen 
![image](https://github.com/user-attachments/assets/1f2f5fc2-8907-418e-a49e-def7ddc580a0)

**Reply by tankerkiller125** [reply]
Let me look into this further, I'm wondering if something isn't quite right with "regional" languages and the way their handled, while the "global" versions are working correctly. (Basically 4 letter with dash language settings vs 2 letter)

**Reply by Jackxwb** [reply]
> It should switch automatically based on your browser settings. You may need to clear the browser cache or CDN for your Homebox domain given it is a JS change. Of course this only applies if your language is supported (see the translations list in the Readme)

I have found the reason, it is that the translation file name does not match navigator.language, causing it to not be recognized
The translation file for Simplified Chinese is "zh_Hans", but navigator. language="zh-CN", so it does not match
![image](https://github.com/user-attachments/assets/9f96cc92-afed-42a4-be99-59ac65fcfbfb)
As for matching traditional Chinese, it may be more complicated. There may be situations such as "zh-HK", "zh-TW", "zh-MO", etc

Perhaps providing a manual language switching option would be more convenient

**Reply by tankerkiller125** [reply]
I've pushed out v0.14.1 with a fix for this issue.

The reason for using the browser instead of a user preference comes down to some technical stuff, but basically getting the user preferences into a Nuxt plugin which is what the localizations needed to work (at least as we currently do it) is deeply challenging as Nuxt plugins are more designed (or so it seems from what I've been able to find so far) to be compiled without application specific functions/services. Which means that while I can change the language on the fly (and I had a MVP of it at one point) I couldn't get it to continue those preferences into the next browser session. 

Hopefully I can figure that out in the future. The compromise of always using the default browser language for now seems like the better choice rather than having a user set their language on every session.

**Reply by Jackxwb** [reply]
> I've pushed out v0.14.1 with a fix for this issue.
> 
> The reason for using the browser instead of a user preference comes down to some technical stuff, but basically getting the user preferences into a Nuxt plugin which is what the localizations needed to work (at least as we currently do it) is deeply challenging as Nuxt plugins are more designed (or so it seems from what I've been able to find so far) to be compiled without application specific functions/services. Which means that while I can change the language on the fly (and I had a MVP of it at one point) I couldn't get it to continue those preferences into the next browser session.
> 
> Hopefully I can figure that out in the future. The compromise of always using the default browser language for now seems like the better choice rather than having a user set their language on every session.

In fact, the user's first entry is on the login page. When the user first visits, I think the option to switch languages can be placed on this page. When the user logs in successfully, they can directly return to the user's personalized settings, and then selectively overwrite the browser cookie or local storage or Vue store before jumping to the homepage. When the homepage page loads, the user's localized configuration can be obtained

**Reply by Jackxwb** [reply]
> I've pushed out v0.14.1 with a fix for this issue.
> 
> The reason for using the browser instead of a user preference comes down to some technical stuff, but basically getting the user preferences into a Nuxt plugin which is what the localizations needed to work (at least as we currently do it) is deeply challenging as Nuxt plugins are more designed (or so it seems from what I've been able to find so far) to be compiled without application specific functions/services. Which means that while I can change the language on the fly (and I had a MVP of it at one point) I couldn't get it to continue those preferences into the next browser session.
> 
> Hopefully I can figure that out in the future. The compromise of always using the default browser language for now seems like the better choice rather than having a user set their language on every session.

V0.14.1 failed to modify the issue
I see that you used the method of replacing "-" to match, but in Chinese Mainland, the browser returns "zh-CN", not "Hans" or "Hant" related fields
What is displayed after "-" is the region, such as "HK" in Hong Kong China, "TW" in Taiwan China, and "MO" in Macao China

As far as I know, simplified Chinese is generally used in Chinese Mainland, while traditional Chinese is widely used in other regions of China

**Reply by Jackxwb** [reply]
> I've pushed out v0.14.1 with a fix for this issue.
> 
> The reason for using the browser instead of a user preference comes down to some technical stuff, but basically getting the user preferences into a Nuxt plugin which is what the localizations needed to work (at least as we currently do it) is deeply challenging as Nuxt plugins are more designed (or so it seems from what I've been able to find so far) to be compiled without application specific functions/services. Which means that while I can change the language on the fly (and I had a MVP of it at one point) I couldn't get it to continue those preferences into the next browser session.
> 
> Hopefully I can figure that out in the future. The compromise of always using the default browser language for now seems like the better choice rather than having a user set their language on every session.

The replacement seems to have caused errors in subsequent methods: Incorrect local information provided
![image](https://github.com/user-attachments/assets/cedea2cb-c6cb-43e8-bba3-8ec1255cea9e)

**Reply by tankerkiller125** [reply]
Can you confirm what `navigator.languages` spits out in your browser? I've created a refactor (which will be in nightly shortly enough) that compares all of the browser languages to what's available, which through some testing on my side works based on some example arrays I was able to find, but I would like some confirmation on what actually gets returned for a real user before tagging a patch release again.

**Reply by Jackxwb** [reply]
> Can you confirm what `navigator.languages` spits out in your browser? I've created a refactor (which will be in nightly shortly enough) that compares all of the browser languages to what's available, which through some testing on my side works based on some example arrays I was able to find, but I would like some confirmation on what actually gets returned for a real user before tagging a patch release again.

`navigator.languages` return `["zh-CN", "en-US"]` In my browser (mobile browser) now

I couldn't find the complete array list, I can only provide some lists that I know
use `zh-hans` array
`["zh-CN"]`
use `zh-hant` array
`["zh-HK","zh-MO","zh-TW"]`

On my way to search for information, I came across a description that says "[Safari country codes on iOS devices smaller than 10.2 are returned in lowercase!" "en-us", "fr-fr", etc. ](https://developer.mozilla.org/zh-TW/docs/Web/API/Navigator/language)"Perhaps we should consider comparing them all in uppercase or lowercase for better compatibility

**Reply by tankerkiller125** [reply]
Why am I not surprised that Apple once again "Thinks Different" and decided not to follow the standards that everyone else follows?

I'll fix that as well, I think it may be better to simply clone the existing Chinese language files to their region specific codes, instead of creating special exceptions in code (because exceptions will start adding up fast if we didn't for every language)

**Reply by Jackxwb** [reply]
> Why am I not surprised that Apple once again "Thinks Different" and decided not to follow the standards that everyone else follows?
> 
> I'll fix that as well, I think it may be better to simply clone the existing Chinese language files to their region specific codes, instead of creating special exceptions in code (because exceptions will start adding up fast if we didn't for every language)

![image](https://github.com/user-attachments/assets/46cc022e-a127-47c4-95a9-cd53510c8645)
I saw on 'sysadminsmedia' that the translation file name related to 'zh-MO' is `Chinese (simplified) (zh_Mo)`. As far as I know, the official promotion in `Macau, China` is `traditional Chinese`. (Check if the translation file content on sysadminsmedia is correct, but if the file name or description is written incorrectly)
According to other descriptions, I believe its name should be: `Chinese (traditional, Macau)`

**Reply by tankerkiller125** [reply]
Translation names aren't really something we control there, that's a weblate thing.

Also please test the nightly build for language, https://nightly.homebox.software/ we also added a language preference option that can be selected in the profile.

**Reply by Jackxwb** [reply]
> Translation names aren't really something we control there, that's a weblate thing.
> 
> Also please test the nightly build for language, https://nightly.homebox.software/ we also added a language preference option that can be selected in the profile.

It can recognize normally! It can recognize `zh-CN` normally on both my computer and phone👌

Highly praised for having language switching function!💕

**Comment by drthanwho** [comment]
I see no one has contributed for Greek, I might as well start on that. Question I have though should it be possible to add items and location in Greek? I cans see the items and all just fine, but I'm unable to search for them correctly. Is that a different kind of limitation I suppose? 

For example right entry written in Greek
<img width="653" height="508" alt="image" src="https://github.com/user-attachments/assets/d96e7ad1-cf9e-4b2f-9bba-e3c0b42a6be0" />
but doesn't come up if you search
<img width="732" height="289" alt="image" src="https://github.com/user-attachments/assets/04c2bdb4-07fd-443f-8db9-ad3e7362ee5d" />

**Reply by tankerkiller125** [reply]
SQLite unfortunately does not handle international search very well. Something we should really document and push a recommendation for Postgres for.
